pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -607,6 +619,80 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 	return &status, nil
 }
 
+func updateStatefulSetAfterInvariantEstablished(
+	ctx context.Context,
+	ssc *defaultStatefulSetControl,
+	set *apps.StatefulSet,
+	replicas []*v1.Pod,
+	updateRevision *apps.ControllerRevision,
+	status apps.StatefulSetStatus,
+) (*apps.StatefulSetStatus, error) {
+
+	replicaCount := int(*set.Spec.Replicas)
+
+	// we compute the minimum ordinal of the target sequence for a destructive update based on the strategy.
+	updateMin := 0
+	maxUnavailable := 1
+	if set.Spec.UpdateStrategy.RollingUpdate != nil {
+		updateMin = int(*set.Spec.UpdateStrategy.RollingUpdate.Partition)
```

Comment:
`Partition` could be nil?

Reply:
Maybe I don't understand how this is supposed to work but it seems to me that this doesn't work if this is a non-zero value?

Reply:
this is all existing old code 
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L578-L583

- partition is defaulted to 0 always , and hence that would be reason its not nil
- if partition is non zero, the pods are deleted starting from highest ordinal to the ordinal equal to the partition number, all pods less than partition are left untouched
- eg. maxUnavailable =2, total replicas 6 partition=3, so the pods deleted are 5,4(because maxUnavailable) first and eventually 3 also. 2, 1 and 0 remain untouched

Reply:
In that case you need to decide if maxUnavailable is per-partition or total, state that in the documentation, and probably fix the logic below.

Reply:
maxUnavailable is total , we dont want to proceed with any update, if the total count of unavailable pods within 0 to replicas is greater than or equal to maxUnavailable